src/basic/chase-symlinks.c

```diff
@@ -113,69 +114,74 @@ int chase_symlinks(
-         *    directory. Note that the returned O_PATH file descriptors must be converted into a regular one (using
-         *    fd_reopen() or such) before it can be used for reading/writing. ret_fd may not be combined with
-         *    CHASE_NONEXISTENT.
+         *    directory. Note that the returned O_PATH file descriptors must be converted into a regular one
+         *    (using fd_reopen() or such) before it can be used for reading/writing. ret_fd may not be
+         *    combined with CHASE_NONEXISTENT.
          *
-         * 3. With CHASE_STEP: in this case only a single step of the normalization is executed, i.e. only the first
-         *    symlink or ".." component of the path is resolved, and the resulting path is returned. This is useful if
-         *    a caller wants to trace the path through the file system verbosely. Returns < 0 on error, > 0 if the
-         *    path is fully normalized, and == 0 for each normalization step. This may be combined with
-         *    CHASE_NONEXISTENT, in which case 1 is returned when a component is not found.
+         * 3. With CHASE_STEP: in this case only a single step of the normalization is executed, i.e. only
+         *    the first symlink or ".." component of the path is resolved, and the resulting path is
+         *    returned. This is useful if a caller wants to trace the path through the file system verbosely.
+         *    Returns < 0 on error, > 0 if the path is fully normalized, and == 0 for each normalization
+         *    step. This may be combined with CHASE_NONEXISTENT, in which case 1 is returned when a component
+         *    is not found.
          *
-         * 4. With CHASE_SAFE: in this case the path must not contain unsafe transitions, i.e. transitions from
-         *    unprivileged to privileged files or directories. In such cases the return value is -ENOLINK. If
-         *    CHASE_WARN is also set, a warning describing the unsafe transition is emitted.
+         * 4. With CHASE_SAFE: in this case the path must not contain unsafe transitions, i.e. transitions
+         *    from unprivileged to privileged files or directories. In such cases the return value is
+         *    -ENOLINK. If CHASE_WARN is also set, a warning describing the unsafe transition is emitted.
          *
          * 5. With CHASE_NO_AUTOFS: in this case if an autofs mount point is encountered, path normalization
          *    is aborted and -EREMOTE is returned. If CHASE_WARN is also set, a warning showing the path of
          *    the mount point is emitted.
          */
 
-        /* A root directory of "/" or "" is identical to none */
-        if (empty_or_root(original_root))
-                original_root = NULL;
-
-        if (!original_root && !ret_path && !(flags & (CHASE_NONEXISTENT|CHASE_NO_AUTOFS|CHASE_SAFE|CHASE_STEP)) && ret_fd) {
+        if (dir_fd < 0 && !ret_path && !(flags & (CHASE_NONEXISTENT|CHASE_NO_AUTOFS|CHASE_SAFE|CHASE_STEP)) && ret_fd) {
```

Comment:
so what's the goal here? you test dir_fd < 0, but then pass that to openat(), which will cause an error if you pass anything but AT_FDCWD.. but if you want to only do that on AT_FDCWD, then maybe write this as `dir_fd == AT_FDCWD`? and in that case you don't need openat() anyway, but can stick to open()

Reply:
I was trying to follow `openat()` behavior here. There's the following snippets in its man page:

> If the pathname given in pathname is absolute, then dirfd is ignored.

> If  the  pathname  given in pathname is relative, and dirfd is not a valid file descriptor, an error (EBADF) results. (Specifying an invalid file descriptor number in dirfd can be used as a means to ensure that pathname is absolute.)

So -1 is actually a valid value for `dir_fd`, as long as the path is absolute. But thinking about it, there's no real reason to support that in `chase_symlinks_at()`, since users can easily just pass `AT_FDCWD` with an absolute path to get exactly the same behavior, so we can enforce that `dir_fd` is either positive or `AT_FDCWD`.